### PR TITLE
Expose perplexity parameter in Seurat

### PR DESCRIPTION
**Pull request by mtekman**
FOR CONTRIBUTOR:
* [ ] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [ ] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [X] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

Related: https://github.com/galaxyproject/tools-iuc/issues/3115

### CODE REVIEW

File: tools/seurat/seurat.xml [hunk 1]
```diff
@@ -69,6 +70,7 @@ Rscript -e "library(\"rmarkdown\"); render(\"$__tool_directory__/Seurat.R\",
             <param name="high_thresholds" type="integer" value="20000000" label="High threshold for filtering cells" />
             <param name="cells_use" type="integer" min="1" value="500" label="Cells to use for PCHeatmap" help="Plots this number of top ‘extreme’ cells on both ends of the spectrum, which dramatically speeds plotting for large datasets" />
             <param name="resolution" type="float" value="0.6" label="Resolution parameter" help="Value of the resolution parameter used in FindClusters, use a value above (below) 1.0 if you want to obtain a larger (smaller) number of communities." />
+            <param name="perplexity" type="integer" value="-1" label="Perplexity parameter" help="Parameter for tSNE plots. Leave at -1 to use recommended defaults." />
```
Comment (nsoranzo) [inline tools/seurat/seurat.xml:73]: I'd suggest:

```suggestion
            <param name="perplexity" type="integer" value="" optional="true" label="Perplexity parameter" help="Parameter for the tSNE dimensionality reduction" />
```

and add:
```
#if not str($adv.perplexity):
    #set $adv.perplexity = -1
#end if
```

to the command.

**Comment by mtekman** [comment]
thanks @nsoranzo

File: tools/seurat/seurat.xml (hunk 1 above)
Comment (nsoranzo) [reply tools/seurat/seurat.xml]: Can you also add the other bit to the command please?

**Comment by mtekman** [comment]
sorry, I must have pushed without pulling

**Comment by nsoranzo** [comment]
It seems you started off an old version of the master branch which doesn't include the changes of https://github.com/galaxyproject/tools-iuc/pull/3057 , can you rebase the PR?

**Comment by mtekman** [comment]
sure

**Comment by mtekman** [comment]
(did not mean to force push that second time)

File: tools/seurat/seurat.xml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <tool id="seurat" name="Seurat" version="@TOOL_VERSION@+galaxy1">
     <description>- toolkit for exploration of single-cell RNA-seq data</description>
     <macros>
-        <token name="@TOOL_VERSION@">3.1.5</token>
+        <token name="@TOOL_VERSION@">3.1.6</token>
```
Comment (nsoranzo) [inline tools/seurat/seurat.xml:4]: 3.1.6 is not yet available on conda: https://anaconda.org/conda-forge/r-seurat

Comment (mtekman) [reply tools/seurat/seurat.xml:4]: oh I thought this was the galaxy version (I was on the old master without the `+galaxy1`)

**Comment by nsoranzo** [comment]
Can you please add:
```
#if not str($adv.perplexity):
    #set $adv.perplexity = -1
#end if
```
inside the `<command>` to deal with an empty value.

**Comment by mtekman** [comment]
apologies, my work machine is preoccupied with another conda so I am eyeballing these edits

**Comment by mtekman** [comment]
So the tests fail, but I think it's because the output html is drastically different than the one before.

The `#+echo` commands that lintr complained about do not show in the final HTML, but also many plots do not exist in the old html

[test1only.zip](https://github.com/galaxyproject/tools-iuc/files/4893180/test1only.zip)

is the lack of violin plots and heatmaps in the original HTML file expected @nsoranzo ?

**Comment by mtekman** [comment]
I don't want to generate a new `out.html` without first being sure it is correct to do

**Comment by bgruening** [comment]
@mtekman I see this here:

```
|  Error in knit_params_get(input_lines, params) :
|  render params not declared in YAML: perplexity
|  Calls: render -> knit_params_get
|  Execution halted
```

**Comment by nsoranzo** [comment]
@bgruening That was before the last commit, right?

**Comment by astrovsky01** [comment]
@mtekman The lack of certain plots was intentional, only one was requested. If you want more plots, add their types to the value for "meta.plots". 

@bgruening,  the kitting has the error because the knit value is being set from a value that doesn't exist in the xml: 
https://github.com/galaxyproject/tools-iuc/blob/cc66319474bf71f40db8297f04c3c3f569e0c3af/tools/seurat/seurat.xml#L59

**Comment by bgruening** [comment]
@nsoranzo seems so. Now its just the diff in the plot: `Files /tmp/tmp7b22xdxkout.html=779292b but /tmp/tmp8d59g1roout.html=1431763b - compare by size (delta=20000) failed`

**Comment by mtekman** [comment]
> @mtekman The lack of certain plots was intentional, only one was requested. If you want more plots, add their types to the value for "meta.plots".

I'm seeing a lot of plots, just for the first test... check my zip file in the previous comment -- it's a bit strange that they would crop up now (unless I've disabled something)

**Comment by astrovsky01** [comment]
I think it might have to do with you using ${adv_perplexity} instead of just $adv_perplexity in the knitting part of the xml, since 
you're redefining the variable?

**Comment by mtekman** [comment]
Can we skip the linting checks? The "comments" it complains about are actually echoed into the PDF document, and so are necessary 

Screenshot_2020-07-22 Expose perplexity parameter in Seurat by mtekman · Pull Request #3116 · galaxyproject tools-iuc

**Comment by bgruening** [comment]
Try to add `# nolint`? https://rdrr.io/cran/lintr/man/exclude.html

**Comment by bgruening** [comment]
Green! @alexlenail can you please review this one?

**Comment by alexlenail** [comment]
@bgruening I don't actually use Seurat, nor R for that matter, so I wouldn't be the best person to review. Taking a quick look at this PR, from what little I know, I don't see anything wrong.

**Comment by bgruening** [comment]
Thanks all!

**Pull request merged**